examples/h_mountain.py

```diff
@@ -0,0 +1,179 @@
+mesh = Mesh(new_coords)
+
+# sponge function
+W_DG = FunctionSpace(mesh, "DG", 2)
+x, z = SpatialCoordinate(mesh)
+zc = H-20000.
+mubar = 0.3/dt
+mu_top = conditional(z <= zc, 0.0, mubar*sin((pi/2.)*(z-zc)/(H-zc))**2)
+mu = Function(W_DG).interpolate(mu_top)
+fieldlist = ['u', 'rho', 'theta']
+timestepping = TimesteppingParameters(dt=dt, alpha=1.0)
+output = OutputParameters(dirname='h_mountain_smootherz', dumpfreq=30, dumplist=['u'], perturbation_fields=['theta', 'rho'])
+parameters = CompressibleParameters(g=9.80665, cp=1004.)
+diagnostics = Diagnostics(*fieldlist)
+diagnostic_fields = [CourantNumber(), VelocityZ(), HydrostaticImbalance()]
+
+state = State(mesh, vertical_degree=1, horizontal_degree=1,
+              family="CG",
+              sponge_function=mu,
+              hydrostatic=True,
+              timestepping=timestepping,
+              output=output,
+              parameters=parameters,
+              diagnostics=diagnostics,
+              fieldlist=fieldlist,
+              diagnostic_fields=diagnostic_fields)
+
+
+# Off-centring parameter is set to alpha=1.0 for first time step
+# to omit spurious orography waves, then alpha=0.5
+def implicit_first_step():
+    if state.t.dat.data_ro[0] == dt:
+        state.timestepping.alpha = 0.5
+
+state.parameter_update = implicit_first_step
+
+# Initial conditions
+u0 = state.fields("u")
+rho0 = state.fields("rho")
+theta0 = state.fields("theta")
+
+# spaces
+Vu = u0.function_space()
+Vt = theta0.function_space()
+Vr = rho0.function_space()
+
+# Thermodynamic constants required for setting initial conditions
+# and reference profiles
+g = parameters.g
+p_0 = parameters.p_0
+c_p = parameters.cp
+R_d = parameters.R_d
+kappa = parameters.kappa
+
+# Hydrostatic case: Isothermal with T = 250
+Tsurf = 250.
+N = g/sqrt(c_p*Tsurf)
+
+# N^2 = (g/theta)dtheta/dz => dtheta/dz = theta N^2g => theta=theta_0exp(N^2gz)
+thetab = Tsurf*exp(N**2*z/g)
+theta_b = Function(Vt).interpolate(thetab)
+
+# Calculate hydrostatic Pi
+params = {'pc_type': 'fieldsplit',
+          'pc_fieldsplit_type': 'schur',
+          'ksp_type': 'gmres',
+          'ksp_monitor_true_residual': True,
+          'ksp_max_it': 1000,
+          'ksp_gmres_restart': 50,
+          'pc_fieldsplit_schur_fact_type': 'FULL',
+          'pc_fieldsplit_schur_precondition': 'selfp',
+          'fieldsplit_0_ksp_type': 'richardson',
+          'fieldsplit_0_ksp_max_it': 5,
+          'fieldsplit_0_pc_type': 'bjacobi',
+          'fieldsplit_0_sub_pc_type': 'ilu',
+          'fieldsplit_1_ksp_type': 'richardson',
+          'fieldsplit_1_ksp_max_it': 5,
+          "fieldsplit_1_ksp_monitor_true_residual": True,
+          'fieldsplit_1_pc_type': 'bjacobi',
+          'fieldsplit_1_sub_pc_type': 'ilu'}
+Pi = Function(Vr)
+rho_b = Function(Vr)
+compressible_hydrostatic_balance(state, theta_b, rho_b, Pi, top=True, pi_boundary=0.5, params=params)
+
+
+def min(f):
+    fmin = op2.Global(1, [1000], dtype=float)
```

Comment:
OK, this kind of bad code suggests we should implement these callbacks in PyOP2.  Because this is so fragile.  If all the values in `f` are larger than 1000, you get the wrong answer, for example.